src/TaskLayout.ts

```diff
@@ -0,0 +1,72 @@
+/**
+ * Various rendering options for a query.
+ */
+export class LayoutOptions {
+    hideTaskCount: boolean = false;
+    hideBacklinks: boolean = false;
+    hidePriority: boolean = false;
+    hideStartDate: boolean = false;
+    hideScheduledDate: boolean = false;
+    hideDoneDate: boolean = false;
+    hideDueDate: boolean = false;
+    hideRecurrenceRule: boolean = false;
+    hideEditButton: boolean = false;
+    hideUrgency: boolean = true;
+    shortMode: boolean = false;
+}
+
+export type LayoutComponent =
+    | 'description'
+    | 'priority'
+    | 'recurrenceRule'
+    | 'startDate'
+    | 'scheduledDate'
+    | 'dueDate'
+    | 'doneDate'
+    | 'blockLink';
+
+/**
+ * This represents the desired layout of tasks when they are rendered in a given configuration.
+ * The layout is used when flattening the task to a string and when rendering queries, and can be
+ * modified by applying LayoutOptions objects.
+ */
+export class TaskLayout {
+    public defaultLayout: LayoutComponent[] = [
+        'description',
+        'priority',
+        'recurrenceRule',
+        'startDate',
+        'scheduledDate',
+        'dueDate',
+        'doneDate',
+        'blockLink',
+    ];
+    public layoutComponents: LayoutComponent[];
+    public options: LayoutOptions;
+
+    constructor(options?: LayoutOptions, components?: LayoutComponent[]) {
+        if (options) this.options = options;
+        else this.options = new LayoutOptions();
+        if (components) this.layoutComponents = components;
+        else this.layoutComponents = this.defaultLayout;
```

Comment:
I value easy readability of code greatly, and almost all uses of one-line ifs without {}, I find that the structure becomes less clear.

In this particular case, I spent some time looking at these 4 lines and trying to work out what the pattern was in the code.

Eventually I added {} and the hidden pattern immediately became clear, because the `this.options` is aligned. Such patterns really help.

```ts
       if (options) {
            this.options = options;
        } else {
            this.options = new LayoutOptions();
        }
        if (components) {
            this.layoutComponents = components;
        } else {
            this.layoutComponents = this.defaultLayout;
        }
```

Having seen the above, I realised that another option might be to use each of the 2 blocks a one-liner by using a ternary operator.